tests/library/test_pydantic.py

```diff
@@ -0,0 +1,174 @@
+    round_trip_object = validate_string(lm[CAPTURE_KEY], pydantic_model)
+    assert round_trip_object == target_obj
+
+
+def test_simple_model():
+    class Simple(pydantic.BaseModel):
+        my_string: str
+
+    my_obj = Simple(my_string="some string")
+    generate_and_check(my_obj, Simple)
+
+
+def test_model_with_int_list():
+    class MyModel(pydantic.BaseModel):
+        my_list: List[int] = pydantic.Field(default_factory=list)
+
+    my_obj = MyModel(my_list=[1, 2, 3, 4])
+    generate_and_check(my_obj, MyModel)
+
+
+def test_nested_model():
+    class A(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_a_str")
+
+    class B(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_b_str")
+        my_A: A = pydantic.Field(default_factory=A)
+
+    class C(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_c_str")
+        my_B: B = pydantic.Field(default_factory=B)
+
+    my_obj = C(my_str="some other string!")
+    generate_and_check(my_obj, C)
+
+
+@pytest.mark.parametrize("has_A", [True, False])
+def test_model_with_optional(has_A):
+    class A(pydantic.BaseModel):
+        my_str: str = pydantic.Field(default="my_a_str")
+
+    class B(pydantic.BaseModel):
+        b_str: str = pydantic.Field(default="Some string")
+        my_A: Union[A, None] = pydantic.Field(default=None)
+
+    if has_A:
+        my_obj = B(my_A=A(my_str="a long string or two"))
+    else:
+        my_obj = B(b_str="A long b string")
+
+    generate_and_check(my_obj, B)
+
+
+@pytest.mark.parametrize("target_obj", ["hello", 42, False])
+def test_literal(target_obj):
+    model = pydantic.TypeAdapter(Literal["hello", 42, False])
+    generate_and_check(target_obj, model)
+
+
+class TestTuple:
+
+    @pytest.mark.parametrize("target_obj", [(1,), (1, 2), (1, 2, 3, 4, 5)])
+    def test_variadic(self, target_obj):
+        model = pydantic.TypeAdapter(tuple[int, ...])
+        generate_and_check(target_obj, model)
+
+    @pytest.mark.xfail(
+        reason="Underlying guidance.json does not yet support sequences with length specifications"
+    )
+    def test_homogeneous(self):
+        model = pydantic.TypeAdapter(tuple[float, float, float])
+        generate_and_check((3.14, 2.718, 1.41), model)
+
+    @pytest.mark.xfail(
+        reason="Underlying guidance.json does not yet support prefixItems"
+    )
+    def test_heterogeneous(self):
+        model = pydantic.TypeAdapter(tuple[int, bool])
+        generate_and_check((1, True), model)
+
+
+class TestDict:
+    def test_simple(self):
+        model = pydantic.TypeAdapter(dict[str, int])
+        generate_and_check({"hello": 42}, model)
+
+    @pytest.mark.xfail(reason="Json schemas cannot specify non-string keys")
+    def test_non_string_keys_fail(self):
+        model = pydantic.TypeAdapter(dict[int, int])
+        bad_str = '{"one":2}'
+        grammar = gen_json(model.json_schema())
+        with pytest.raises(ParserException):
```

Comment:
I suggest that this (and the similar test below) checks the message (or anything else appropriate about the exception... will a `ParserException` have a `current_byte`?). Something like:

```python
with pytest.raises(ParserException) as pe:
    grammar.match(bad_str, raise_exceptions=True)
assert pe.value.args[0] == "expected exception message"
```

This just makes sure that the exception was raised where you were expecting, and not due to some other failure.

Reply:
Totally agreed :) I'll put in some work on these tests when I have an opportunity this weekend